apis/compute/v1beta1/firewall_types.go

```diff
@@ -0,0 +1,265 @@
+	// because service accounts are associated with an instance, not an IP
+	// address. sourceRanges can be set at the same time as
+	// sourceServiceAccounts. If both are set, the firewall applies to
+	// traffic that has a source IP address within the sourceRanges OR a
+	// source IP that belongs to an instance with service account listed in
+	// sourceServiceAccount. The connection does not need to match both
+	// fields for the firewall to apply. sourceServiceAccounts cannot be
+	// used at the same time as sourceTags or targetTags.
+	// +optional
+	SourceServiceAccounts []string `json:"sourceServiceAccounts,omitempty"`
+
+	// TargetServiceAccounts: A list of service accounts indicating sets of
+	// instances located in the network that may make network connections as
+	// specified in allowed[]. targetServiceAccounts cannot be used at the
+	// same time as targetTags or sourceTags. If neither
+	// targetServiceAccounts nor targetTags are specified, the firewall rule
+	// applies to all instances on the specified network.
+	// +optional
+	TargetServiceAccounts []string `json:"targetServiceAccounts,omitempty"`
+
+	// Allowed: The list of ALLOW rules specified by this firewall. Each
+	// rule specifies a protocol and port-range tuple that describes a
+	// permitted connection.
+	// +optional
+	Allowed []*FirewallAllowed `json:"allowed,omitempty"`
+
+	// Denied: The list of DENY rules specified by this firewall. Each rule
+	// specifies a protocol and port-range tuple that describes a denied
+	// connection.
+	// +optional
+	Denied []*FirewallDenied `json:"denied,omitempty"`
+
+	// Direction: Direction of traffic to which this firewall applies,
+	// either `INGRESS` or `EGRESS`. The default is `INGRESS`. For `INGRESS`
+	// traffic, you cannot specify the destinationRanges field, and for
+	// `EGRESS` traffic, you cannot specify the sourceRanges or sourceTags
+	// fields.
+	//
+	// Possible values:
+	//   "EGRESS"
+	//   "INGRESS"
+	// +optional
+	Direction *string `json:"direction,omitempty"`
+
+	// Disabled: Denotes whether the firewall rule is disabled. When set to
+	// true, the firewall rule is not enforced and the network behaves as if
+	// it did not exist. If this is unspecified, the firewall rule will be
+	// enabled.
+	// +optional
+	Disabled *bool `json:"disabled,omitempty"`
+
+	// LogConfig: This field denotes the logging options for a particular
+	// firewall rule. If logging is enabled, logs will be exported to
+	// Stackdriver.
+	// +optional
+	LogConfig *FirewallLogConfig `json:"logConfig,omitempty"`
+}
+
+// FirewallAllowed represents the ALLOW rule by the firewall
+type FirewallAllowed struct {
+	// IPProtocol: The IP protocol to which this rule applies. The protocol
+	// type is required when creating a firewall rule. This value can either
+	// be one of the following well known protocol strings (tcp, udp, icmp,
+	// esp, ah, ipip, sctp) or the IP protocol number.
+	IPProtocol string `json:"IPProtocol,omitempty"`
```

Comment:
Should this be required within the embedded struct? If so it should have `omitempty` removed.